client/network/src/block_request_handler.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Helper for handling (i.e. answering) block requests from a remote peer via the
+//! [`crate::request_responses::RequestResponsesBehaviour`].
+
+use codec::{Encode, Decode};
+use crate::chain::Client;
+use crate::config::ProtocolId;
+use crate::protocol::{message::BlockAttributes};
+use crate::request_responses::{IncomingRequest, ProtocolConfig};
+use crate::schema::v1::block_request::FromBlock;
+use crate::schema::v1::{BlockResponse, Direction};
+use futures::channel::{mpsc, oneshot};
+use futures::stream::StreamExt;
+use log::debug;
+use prost::Message;
+use sp_runtime::generic::BlockId;
+use sp_runtime::traits::{Block as BlockT, Header, One, Zero};
+use std::cmp::min;
+use std::sync::{Arc};
+use std::time::Duration;
+
+const LOG_TARGET: &str = "block-request-handler";
+const MAX_BLOCKS_IN_RESPONSE: usize = 128;
+const MAX_BODY_BYTES: usize = 8 * 1024 * 1024;
+
+/// Generates a [`ProtocolConfig`] for the block request protocol, refusing incoming requests.
+pub fn generate_protocol_config(protocol_id: ProtocolId) -> ProtocolConfig {
+	ProtocolConfig {
+		name: generate_protocol_name(protocol_id).into(),
+		max_request_size: 1024 * 1024,
+		max_response_size: 16 * 1024 * 1024,
+		request_timeout: Duration::from_secs(40),
+		inbound_queue: None,
```

Comment:
Why not have this function take a `mpsc::Receiver<IncomingRequest>`, given that we're providing one on line `74`? I feel like this change would make the code easier to read, instead of someone being like 'Oh there's no `inbound_queue` for this protocol. Oh wait, nevermind.'.

Reply:
I think the answer lies in the fact that this is a "default configuration" and a public function and it is indeed permissible to use a configuration that is not capable of receiving requests. You can find the following comment in other parts of the code:
```rust
	/// Can be constructed either via [`block_request_handler::generate_protocol_config`] allowing
	/// outgoing but not incoming requests, or constructed via
	/// [`block_request_handler::BlockRequestHandler::new`] allowing both outgoing and incoming
	/// requests.
        pub block_request_protocol_config: RequestResponseConfig,
```
I guess you're also not supposed to pass in your own `mpsc::Receiver`, so either you don't support inbound requests or you must use `BlockRequestHandler::new`, hence even taking an `Option<mpsc::Receiver<IncomingRequest>>` may not be desirable for the public API of this module. Just an educated guess.

Reply:
To extend Roman's comment above ...

We have to support two use-cases:

1. Supporting outgoing requests as well as incoming requests, e.g. what a full-node would do.

2. Supporting outgoing requests only, e.g. what a light-client would do.

As you correctly noted above, one differentiates between the two by specifying an `inbound_queue` or not (see `ProtocolConfig` [doc](https://github.com/paritytech/substrate/blob/f74de63d23214b530ac24ac3ea4ead1180e22e83/client/network/src/request_responses.rs#L81-L101)).

A `ProtocolConfig` for the first use-case (full-node) can be created via `BlockRequestHandler::new`. I decided to have `BlockRequest::new` return both the `BlockRequestHandler` and the `ProtocolConfig` to force callers to create both, i.e. not create a `ProtocolConfig` with an `inbound_queue` without a handler listening on the other side of that queue. With that in mind, and as Roman already explained above, taking a `Option<mpsc::Receiver<IncomingRequest>>` in `generate_protocol_config` would allow callers to misuse the API.

A `ProtocolConfig` for the second use-case (light-client) can be created via `generate_protocol_config`.

This is not cast in stone. I am happy for alternative suggestions. From my point of view it is confusing to create a block request `ProtocolConfig` via a function in `block_request_handler.rs ` even in the case where one explicity does not want to handle incoming requests (light-client).